src/ess/reduce/ui.py

```diff
@@ -232,3 +234,136 @@ def refresh_workflow_box(change) -> None:
     workflow_selection_box = widgets.HBox([workflow_select], layout=default_layout)
     workflow_box = widgets.Box(layout=default_layout)
     return widgets.VBox([workflow_selection_box, workflow_box])
+
+
+def _has_widget_value_setter(widget: widgets.Widget) -> bool:
+    widget_type = type(widget)
+    return (
+        widget_property := getattr(widget_type, 'value', None)
+    ) is not None and getattr(widget_property, 'fset', None) is not None
+
+
+def _get_parameter_box(widget: WorkflowWidget | ParameterBox) -> ParameterBox:
+    if isinstance(widget, WorkflowWidget):
+        return widget.parameter_box
+    elif isinstance(widget, ParameterBox):
+        return widget
+    else:
+        raise TypeError(
+            f"Expected target_widget to be a WorkflowWidget or ParameterBox, "
+            f"got {type(widget)}."
+        )
+
+
+def set_input_widget_values(
+    widget: WorkflowWidget | ParameterBox,
+    new_parameter_values: dict[str | type, dict[str, Any]],
+) -> None:
+    """Set the values of the input widgets in the target widget.
+
+    Example
+    -------
+    {
+        'WavelengthBins': {
+            'fields': {'start': 1.0, 'stop': 14.0, 'nbins': 500}
+        }
+    }
+
+    Parameters
+    ----------
+    widget:
+        The widget containing the input widgets.
+    new_parameter_values:
+        A dictionary of parameter values to set for the input widgets.
+        The keys are the parameter names or the type(Key in Sciline graphs).
+        The values are dictionaries with the following keys
+
+        - 'fields':
+            A dictionary of field names and values to set.
+            i.e. {'dim': 'tof', 'unit': 'mm'}
+            The field values will be set if the field exists in the widget.
+            Otherwise, the field value will be ignored and a warning will be raised.
+        - 'value':
+            If the widget has a value setter, the value will be set directly.
+            Otherwise, it will be ignored and a warning will be raised.
```

Comment:
> Do I understand correctly that it is always either-or?

I guess...? But I wasn't entirely sure. We can assume they are.

> or are there other requirements I am not aware of? If it is about storing the "optional" flags etc., maybe it should be

Yes, we'll have to store optional/switchable widget status as well. I'll go for the second option.

Reply:
If the value is a dictionary, isn't that the same as a widget having multiple fields?

Reply:
Not necessarily, a widget can decide how to make the `value` from `fields`.

For example, like this:

```python

class ROIWidget(GridBox, ValueWidget):
    def __init__(self):
        super().__init__()

        self.fields = {
            'xmin': FloatText(description='xmin'),
            'xmax': FloatText(description='xmax'),
            'ymin': FloatText(description='ymin'),
            'ymax': FloatText(description='ymax'),
            'unit': Text(description='unit'),
        }
        self.children = [
            Label(value="Select ROI:"),
            self.fields['xmin'],
            self.fields['xmax'],
            self.fields['ymin'],
            self.fields['ymax'],
            self.fields['unit'],
        ]

    @property
    def value(self):
        return {
            'xmin': sc.scalar(self.fields['xmin'].value, unit=self.fields['unit']),
            'xmax': sc.scalar(self.fields['xmax'].value, unit=self.fields['unit']),
            'ymin': sc.scalar(self.fields['ymin'].value, unit=self.fields['unit']),
            'ymax': sc.scalar(self.fields['ymax'].value, unit=self.fields['unit']),
        }
```

Reply:
This one is an example I just made up btw

Reply:
1. Can we focus on our widgets and not on hypothetical situations?
2. We should distinguish between what the widget returns to the workflow (the `value`, your dict above) and the values of the fields. They don't need to be the same, so I am not sure what you example is demonstrating. Configuring the workflow would use `widget.value`. Saving/restoring the fields could use something like `widget.get_fields` and `widget.set_fields`. Using `widget.value` to get the field values isn't an option anyway, for widgets like the bin-edge widget, so whether there is a widget where `widget.value` returns a `dict` should be a moot point.

Reply:
Are you suggesting to write `sc.vector([1,2,3], unit='m')`  in a text input widget and use `eval` on that?

Probably not, let us talk about this in person tomorrow or so.

Reply:
> or are there other requirements I am not aware of? If it is about storing the "optional" flags etc., maybe it should be

I have an idea how to handle this in wrapping widgets.
Just leaving a note so that I don't forget...